scripts/check_frontend_test_coverage.py

```diff
@@ -0,0 +1,217 @@
+        tests_array = f.read().split('end_of_record')
+        tests_array_filtered = []
+
+        for test in tests_array:
+            lines = [line for line in test.splitlines() if filter_lines(line)]
+            if len(lines) > 0:
+                tests_array_filtered.append(lines)
+
+        return tests_array_filtered
+
+
+def build_tests_fully_coverage_dict():
+    """Build a dict with only fully covered files from develop branch.
+
+    Returns:
+        Dictionary. A dict containing file path, total lines and covered lines
+        of each tested file.
+
+    Raises:
+      Exception: If DEV_LCOV_FILE_PATH doesn't exist.
+    """
+    coverage_dict = {}
+
+    if os.path.exists(DEV_LCOV_FILE_PATH):
+        tests = get_lcov_file_tests(DEV_LCOV_FILE_PATH)
+        for lines in tests:
+            total_lines = lines[1].split(':')[1]
+            covered_lines = lines[2].split(':')[1]
+
+            if total_lines == covered_lines:
+                test_name = get_test_file_name(lines[0])
+                coverage_dict[test_name] = [
+                    int(total_lines),
+                    int(covered_lines)
+                ]
+
+        return coverage_dict
+    else:
+        raise Exception(
+            'File at path {} doesn\'t exist'.format(DEV_LCOV_FILE_PATH))
+
+
+def lcov_files_diff():
+    """Check if any 100% covered file had the coverage dropped.
+
+    Raises:
+      Exception: If PR_LCOV_FILE_PATH doesn't exist.
+    """
+    fully_covered_tests = build_tests_fully_coverage_dict()
+
+    if os.path.exists(PR_LCOV_FILE_PATH):
+        tests = get_lcov_file_tests(PR_LCOV_FILE_PATH)
+        for lines in tests:
+            test_name = get_test_file_name(lines[0])
+
+            if test_name in fully_covered_tests:
+                total_lines = lines[1].split(':')[1]
+                covered_lines = lines[2].split(':')[1]
+
+                test = fully_covered_tests[test_name]
+                total_lines_test = test[0]
+
+                if (total_lines_test == int(total_lines)
+                        and int(total_lines) != int(covered_lines)):
+                    sys.stderr.write(
+                        'The {} file is fully covered and it has '
+                        'decreased after the changes \n'.format(test_name))
+                    sys.exit(1)
+    else:
+        raise Exception('File at path {} doesn\'t exist'.format(
+            PR_LCOV_FILE_PATH))
+
+
+def main():
+    """The first function to be executated."""
+    current_branch = subprocess.check_output([
+        'git', 'rev-parse', '--abbrev-ref', 'HEAD']).strip()
+
+    if current_branch != 'develop' and current_branch != 'master':
+        run_frontend_tests_script()
+
+        create_tmp_folder()
+
+        shutil.copyfile(
+            os.path.join(*LCOV_FILE_PATH.split('/')),
+            os.path.join(*PR_LCOV_FILE_PATH.split('/')))
+
+        change_git_branch('develop')
+
+        run_frontend_tests_script()
```

Comment:
This construction isn't ideal since you're running the frontend tests twice. Is there a way we can store some of these reports in file storage service?

Reply:
That being said, I don't mind this being a short term solution. I just don't want to be stuck doing this for the long term.

Reply:
I think the ideal approach for this is to store the reports from the develop branch in the file storage service after a new merge. Can you tell me more about the file storage service? Where can I find it in the project?

Reply:
We actually don't have any file storage solutions at this point, but may be consider this with @nithusha21.

Reply:
Should we have a checked in file which gets auto-updated when frontend tests run? 

This file would have the latest lcov report for the develop branch. This could probably cause some confusion while reviewing. But that's the only thing I can think of 😅. We'll discuss this in our meeting and get back to you.

Reply:
Hmm, that's probably not a good idea for the exact reason that you've mentioned. It's best to mitigate this, but if we're looking at running this only for the next couple of months then it's okay.

Reply:
If I understand correctly, this function is running the tests on a PR, then running the tests on develop, and comparing the coverage of both. So, it seems to me like we would need to run the frontend tests twice, since these would effectively be different codebases? Please correct me if I'm misunderstanding this.

Also, I'm wondering why you need to use files at all. Couldn't you store the reports in local variables for the purposes of this check?

Reply:
@seanlip That's exactly what the script is doing. I'm using files because that's how karma and istanbul reports the coverage. I don't think there is an option to store the reports in a variable, just in files like .html, .lcov, etc. Lcov was the file type that I chosenbecause it is used on codecov and it has clearly the test total lines and covered lines.

Reply:
Thanks for explaining @marianazangrossi. In that case, I'm a bit confused by @kevinlee12's comment about why running run_frontend_tests() twice is suboptimal. @kevinlee12 could you please clarify your thinking here?

And yeah, I think you can just use the local file system, just store the stuff in a temporary folder and delete it after use. I believe we already have a karma-coverage-reports folder that gets generated similarly. I wouldn't have a checked-in file though as @nithusha21 suggested in his comment, it would cause confusion IMO because it's hard to tell, as the code changes, what exactly that file is referring to.

Reply:
The reason for it being suboptimal is that we'd have to run the coverage for every single PR for develop in the queue, which will add up over time over a couple of PRs and take up our container quota as well (since we're only limited to 4). That being said, if it's only for a limited period, then that's fine.

Reply:
Ideally this check should be removed, and we should revert to a codecov check. This is for the transient period when we would like to enforce 100% for the files already covered, so that we won't need to go back and forth.

Reply:
I talked to @kevinlee12 about this and here's what we came up with: let's not do a comparison with develop. It seems wasteful to re-calculate that for every PR that makes it into the (same) develop branch. 

Instead, let's add a whitelist to the repository -- this could just be a python variable with a list of filepaths. As a filepath reaches 100% coverage, it is added to the list. If a file is renamed, then tests will fail (ideally at presubmit stage) and the contributor would be asked to fix the list accordingly; such changes will be clearly visible to the reviewer who can therefore see at a glance whether the coverage has increased/decreased. The benefit of this is that we're using the codebase itself as the "file system" here, since the list of fully-covered files is stored in the code itself.

In addition, one thing we can do is throw an error in the presubmit/lint script if the frontend tests run and we see a 100%-coverage frontend file that isn't in the list. The error message would then ask the user to add that file to the list.

I think this should be foolproof. What do you think @marianazangrossi -- would this work?

Reply:
I think it would work., it's way better than the approach I've created. I just have some questions:

The script will run before the push, right?

Let's suppose there is a 100% coverage file that isn't in the whitelist. Should the user add this file manually? Or is the script going to insert it automatically?

Reply:
Yes, that is correct. It should run with the usual pre-push checks (that run the linter + Karma frontend tests).

In the scenario you describe, the user should add the file manually, and the script should report any mismatches. The script should not be changing files. The whitelist should have clear instructions for developers about what should be put in there.

Reply:
(And we should also inform oppia-dev@ once this policy is in place, so that reviewers are aware.)

Reply:
Hi @seanlip, just an update. We discussed this yesterday. She was a bit confused as to what you meant. I clarified it for her, and she said she will go ahead and implement it!

Reply:
I started to implement it but I stuck trying to understand the flow. How can I know the difference between reaching a file to 100% coverage and renaming a file that is 100% coverage? The first one should fail the job and the second one should only tell the user to add the renamed file in the whitelist. But how will I differ these situations?

Reply:
I don't think you need to. Just do the following:

- If a file is in the whitelist and it currently exists but does not have 100% test coverage, tell the user "According to the whitelist, this file should have 100% test coverage; please fix."
- If a file is in the whitelist and it is currently deleted, tell the user "According to the whitelist, this file should have 100% test coverage; but it is not found; please fix."
- If a file is at 100% coverage but is not in the whitelist, tell the user "Files XYZ have 100% frontend test coverage, and should be added to the whitelist. Please fix."

That should suffice. (Feel free to change the wording of the error messages to be more user-friendly, e.g. be clear on where this "whitelist" actually is.)